specification/dartLangSpec.tex

```diff
@@ -692,32 +711,55 @@ \subsection{Scoping}
 if a key of \NamespaceName{} is mapped to $d$.
 
 \LMHash{}%
-A scope $S_0$ induces a namespace \NamespaceName{0} that maps
-the simple name of each variable, type or function declaration $d$
-declared in $S_0$ to $d$.
-Labels are not included in the induced namespace of a scope;
-instead they have their own dedicated namespace.
+A scope $S_0$ has an associated namespace \NamespaceName{0}.
+The bindings of \NamespaceName{0} is specified in this document by saying that
+a given declaration $D$ named $n$
+\IndexCustom{introduces}{declaration!introduces an entity into a scope}
+a specific entity $V$ into $S_0$,
+which means that the binding $n\mapsto{}V$ is added to \NamespaceName{0}.
 
 \commentary{%
-It is therefore impossible, e.g., to define a class that declares
-a method and a getter with the same name in Dart.
-Similarly one cannot declare a top-level function with
-the same name as a library variable or a class
-which is declared in the same library.%
+In some cases, the name of the declaration differs from
+the identifier that occurs in the declaration syntax used to declare it.
+Setters have names that are distinct from the corresponding getters
+because they always have an \lit{=} automatically added at the end,
+and the unary minus operator has the special name unary-.%
+}
+
+\commentary{%
+It is typically the case that $V$ is $D$,
+but there are exceptions:
+A variable declaration introduces an implicitly induced getter declaration,
+and in some cases also an implicitly induced setter declaration into the
+given scope.%
+}
+
+\commentary{%
+Note that labels are not included in the namespace of a scope.
+They are resolved lexically rather then being looked up in a namespace.%
 }
 
 \LMHash{}%
 It is a compile-time error if there is more than one entity with the same name
 declared in the same scope.
 
 \commentary{%
-In some cases, the name of the declaration differs from
-the identifier that occurs in the declaration syntax used to declare it.
-Setters have names that are distinct from the corresponding getters
-because they always have an \lit{=} automatically added at the end,
-and unary minus has the special name unary-.%
+It is therefore impossible, e.g., to define a class that declares
+a method and a getter with the same name in Dart.
+Similarly one cannot declare a top-level function with
+the same name as a library variable or a class
+which is declared in the same library.%
 }
 
+\LMHash{}%
+In order to specify a few properties of execution,
+we introduce the notion of a
+\Index{dynamic namespace}.
+This is a partial function from names to run-time entities,
+in particular objects and compiled code.
+Each dynamic namespace corresponds to a namespace with the same keys,
+but with values that correspond to the semantics of the namespace values.
```

Comment:
Somewhat dense paragraph.

How about calling it an "environment" or "runtime environment" instead of "dynamic namespace". It's not the name-space itself which is dynamic, it's the values.

How about:

_A namespace maps names to the declaration, and can be used statically to figure out what a name or variable refers to. At run-time, a variable has an actual value._
We introduce the notion of a *run-time environment* based on a namespace.
The run-time environment has the same keys as the corresponding namespace, but its values are run-time entities, in particular objects and compiled code.
_During execution, the same code can be executed multiple times with a new run-time environment each time._